File: maximum-difference-between-node-and-ancestor/maximum-difference-between-node-and-ancestor.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

var result int

func maxAncestorDiff(root *TreeNode) int {
    result = 0
    helper(root, root.Val, root.Val)
    return result
}

func helper(root *TreeNode, currMax, currMin int) {
    if root == nil {
        return
    }
    result = max(result, max(abs(currMax - root.Val), abs(currMin - root.Val)))
    helper(root.Left, max(currMax, root.Val), min(currMin, root.Val))
    helper(root.Right, max(currMax, root.Val), min(currMin, root.Val))
}

func abs(a int) int {
    if a < 0 {
        return a * -1
    }
    return a
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

File: maximum-difference-between-node-and-ancestor/maximum-difference-between-node-and-ancestor.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    private int result = 0;
    
    public int maxAncestorDiff(TreeNode root) {
        result = 0;
        maxAncestorDiff(root, root.val, root.val);
        return result;
    }
    
    private void maxAncestorDiff(TreeNode root, int currMax, int currMin) {
        if (root == null) {
            return;
        }
        result = Math.max(result, Math.max(Math.abs(root.val - currMax), Math.abs(root.val - currMin)));
        maxAncestorDiff(root.left, Math.max(currMax, root.val), Math.min(currMin, root.val));
        maxAncestorDiff(root.right, Math.max(currMax, root.val), Math.min(currMin, root.val));
    }
}